src/cpu.hpp

```diff
@@ -0,0 +1,368 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2022 YADRO
+
+#pragma once
+
+#include <core/entity/dbus_query.hpp>
+#include <core/entity/entity.hpp>
+#include <formatters.hpp>
+#include <phosphor-logging/log.hpp>
+#include <status_provider.hpp>
+#include <sensors.hpp>
+
+namespace app
+{
+namespace obmc
+{
+namespace entity
+{
+using namespace app::entity;
+using namespace app::query;
+using namespace phosphor::logging;
+
+class CpuSMBiosProvider final :
+    public entity::EntitySupplementProvider,
+    public ISingleton<CpuSMBiosProvider>,
+    public CachedSource,
+    public NamedEntity<CpuSMBiosProvider>
+{
+  public:
+    enum class ProcessorType
+    {
+        cpu,
+        unknown
+    };
+    enum class InstructionSet
+    {
+        x86_64,
+        powerISA,
+        unknown
+    };
+    enum class Arch
+    {
+        x86,
+        power,
+        unknown
+    };
+
+    ENTITY_DECL_FIELD(std::string, Name)
+    ENTITY_DECL_FIELD_ENUM(ProcessorType, ProcessorType, unknown)
+    ENTITY_DECL_FIELD_ENUM(InstructionSet, InstructionSet, unknown)
+    ENTITY_DECL_FIELD_ENUM(Arch, Arch, unknown)
+
+    ENTITY_DECL_FIELD(std::string, Characteristics)
+    ENTITY_DECL_FIELD(std::string, Family)
+    ENTITY_DECL_FIELD(std::string, Manufacturer)
+    ENTITY_DECL_FIELD(std::string, Socket)
+    ENTITY_DECL_FIELD(std::string, Model)
+
+    ENTITY_DECL_FIELD_DEF(uint16_t, CoreCount, 0)
+    ENTITY_DECL_FIELD_DEF(uint32_t, IdentityReg, 0)
+    ENTITY_DECL_FIELD_DEF(uint16_t, MaxSpeed, 0)
+    ENTITY_DECL_FIELD_DEF(uint16_t, ThreadCount, 0)
+
+  private:
+    class Query final : public dbus::FindObjectDBusQuery
+    {
+        static constexpr const char* cpuService =
+            "xyz.openbmc_project.Smbios.MDR_V2";
```

Comment:
In some cases it is required to specify a target service where we search for objects. In this case we must obtain objects from `mdrv2` because the same objects (as you know) will be found at the dbus-sensors (cpusensors). The `CpuSMBiosProvider` purpose is to provide a supplement payload for cpu-inventory that was created by ProcessorSummary IEntity.

Reply:
Not the same: it provides different interfaces, so you can use more complex conditions to query required one. But probably you are right, we likely should simplify overall system.